### PR TITLE
Use tape based Libtask

### PR DESCRIPTION
**Pull request by KDr2**
Run test cases against the tape-based libtask.

We don't support branches in taped libtask now, so I commented out one test case which contains a `if` expression.

All other test cases passed.

### CODE REVIEW

**Comment by devmotion** [comment]
I think it would be good to make a release of AdvancedPS 0.3 first, update Turing, and add integration tests here. There are many unreleased changes on the master branch of AdvancedPS and it is not clear how well they work with Turing and what changes are needed there. I think it would be better to update the ecosystem in two steps, to avoid mixing up problems or introducing any regressions in the Libtask rewrite (still think it would be better to use a new package :smile:).

File: src/model.jl [hunk 1]
```diff
@@ -6,3 +6,14 @@ Observe sample `x` from distribution `dist` and yield its log-likelihood value.
 function observe(dist::Distributions.Distribution, x)
     return Libtask.produce(Distributions.loglikelihood(dist, x))
 end
+
+function (instr::Libtask.Instruction{typeof(observe)})()
```
Comment (devmotion) [inline src/model.jl:10]: Why is this needed? And could we dispatch on the arguments of `observe` if at some point we decide to implement add more `observe` implementations?

Comment (KDr2) [reply src/model.jl:10]: This is needed because the current tape impl only tracks the direct call to `produce`, the task function calls this function, and this calls `produce`, so `produce` is not directly called by the task function and won't be recorded. This function definition makes `observe` have the same effect as `produce`.

The Instruction type is not parameterized on its input types now, I think it is possible to add this in. I will look into that later.

**Comment by KDr2** [comment]
> still think it would be better to use a new package 😄

I have the same thought, but I am focusing on making sure this approach works now and haven't got any concrete update plans :)

**Comment by yebai** [comment]
> I think it would be good to release AdvancedPS 0.3 first, update Turing, and add integration tests here. There are many unreleased changes on the master branch of AdvancedPS and it is not clear how well they work with Turing and what changes are needed there. 

How about we first make a new release based on the most recent release, but with changes in this PR? This would allow us to support Julia 1.7 first. I think new changes on the master branch and in #32 will likely take more time to push through. 

I have created a new [releases branch](https://github.com/TuringLang/AdvancedPS.jl/tree/releases) from the most recent release. If we merge this PR into the releases brach, we can make a new release there without worrying about the changes on the master branch.

File: src/container.jl [hunk 2]
```diff
@@ -62,13 +56,8 @@ function forkr(trace::Trace)
...
 
-    ctask = let f = trace.ctask.task.code
-        Libtask.CTask() do
-            res = f()(trace.rng)
-            Libtask.produce(nothing)
-            return res
-        end
-    end
+    func = Libtask.func(trace.ctask)
+    ctask = Libtask.CTask(func, trace.rng)
```
Comment (yebai) [inline src/container.jl:60]: Maybe create a convenient constructor `Libtask.CTask(t::Ctask, args...)` so we don't need to handle `func` explicitly.

Comment (KDr2) [reply src/container.jl]: Done.

**Comment by KDr2** [comment]
Close in favor of https://github.com/TuringLang/AdvancedPS.jl/pull/36.

**Comment by yebai** [comment]
#36 is a backport to the most recent release, we still need this PR to update the master.

File: src/model.jl [hunk 3]
```diff
@@ -6,3 +6,5 @@ Observe sample `x` from distribution `dist` and yield its log-likelihood value.
 function observe(dist::Distributions.Distribution, x)
     return Libtask.produce(Distributions.loglikelihood(dist, x))
 end
+
+Libtask.trace_into(::typeof(observe)) = true
```
Comment (yebai) [inline src/model.jl:10]:
```suggestion
```

File: test/container.jl [hunk 4]
```diff
@@ -12,7 +12,7 @@
         function fpc(logp)
             f = let logp = logp
                 rng -> begin
-                    while true
+                    for _ in 1:100
```
Comment (yebai) [inline test/container.jl:15]:
```suggestion
                    while true
```

File: test/container.jl [hunk 5]
```diff
@@ -100,7 +100,7 @@
         function f2(rng)
             t = TArray(Int, 1)
             t[1] = 0
-            while true
+            for _ in 1:100
```
Comment (yebai) [inline test/container.jl:103]:
```suggestion
            while true
```

File: test/smc.jl [hunk 6]
```diff
@@ -55,12 +55,12 @@
         function (m::FailSMCModel)(rng::Random.AbstractRNG)
             m.a = a = rand(rng, Normal(4, 5))
             m.b = b = rand(rng, Normal(a, 1))
-            if a >= 4
-                AdvancedPS.observe(Normal(b, 2), 1.5)
-            end
+            # if a >= 4
```
Comment (yebai) [inline test/smc.jl:58]:
```suggestion
            if a >= 4
```

File: test/smc.jl [hunk 7]
```diff
@@ -55,12 +55,12 @@
...
+            AdvancedPS.observe(Normal(b, 2), 1.5)
+            # end
```
Comment (yebai) [inline test/smc.jl:60]:
```suggestion
            end
```

File: test/smc.jl [hunk 8]
```diff
@@ -55,12 +55,12 @@
...
         end
 
-        @test_throws ErrorException sample(FailSMCModel(), AdvancedPS.SMC(100))
+        # @test_throws ErrorException sample(FailSMCModel(), AdvancedPS.SMC(100))
```
Comment (yebai) [inline test/smc.jl:63]:
```suggestion
        @test_throws ErrorException sample(FailSMCModel(), AdvancedPS.SMC(100))
```

**Review by yebai — CHANGES_REQUESTED** [review]

File: src/container.jl [hunk 9]
```diff
@@ -7,13 +7,7 @@ end
...
 
 function Trace(f, rng::TracedRNG)
-    ctask = let f = f
-        Libtask.CTask() do
-            res = f(rng)
-            Libtask.produce(nothing)
-            return res
-        end
-    end
+    ctask = Libtask.CTask(f, rng)
```
Comment (FredericWantiez) [inline src/container.jl:10]: Why not match the new Turing interface ? Or that's for a later PR ?
```julia
  if hasfield(typeof(f), :evaluator) # Test whether f is a Turing.TracedModel
      # println(f.evaluator)
      ctask = Libtask.CTask(f.evaluator[1], f.evaluator[2:end]...)
  else # f is a Function, or AdavncedPS.Model
      ctask = Libtask.CTask(f)
  end
```

Comment (yebai) [reply src/container.jl:10]: yes, we should match that - please feel free to update this PR!

Comment (devmotion) [reply src/container.jl:10]: Couldn't/Shouldn't this be handled in Turing by implementing `Libtask.CTask(f::TracedModel, rng)`?

File: src/container.jl [hunk 10]
```diff
@@ -62,13 +56,7 @@ function forkr(trace::Trace)
...
     Random123.set_counter!(trace.rng, 1)
 
-    ctask = let f = trace.ctask.task.code
-        Libtask.CTask() do
-            res = f()(trace.rng)
-            Libtask.produce(nothing)
-            return res
-        end
-    end
+    ctask = Libtask.CTask(trace.ctask, trace.rng)
```
Comment (yebai) [inline src/container.jl:59]:
```suggestion
    ctask = Libtask.CTask(newf, trace.rng)
```
```suggestion
    ctask = Libtask.CTask(trace.ctask, trace.rng)
```
